ha/lock_test.go

```diff
@@ -0,0 +1,334 @@
+	ml.mu.Lock()
+	defer ml.mu.Unlock()
+
+	ml.acquiresCalls[callerID] += 1
+	if ml.locked {
+		return "", nil
+	}
+
+	ml.locked = true
+	ml.leaseStartTime = time.Now()
+	return "lockID", nil
+}
+
+func (ml *mockLock) Release(_ context.Context) error {
+	ml.mu.Lock()
+	defer ml.mu.Unlock()
+
+	if !ml.locked {
+		return errors.New("error")
+	}
+
+	ml.locked = false
+	ml.renewsCounter = 0
+	return nil
+}
+
+// The behavior of renew is not an exact replication of
+// the lock work, its intended to test the behavior of the
+// multiple instances running.
+func (ml *mockLock) Renew(_ context.Context) error {
+	ml.mu.Lock()
+	defer ml.mu.Unlock()
+
+	if !ml.locked {
+		return errors.New("error")
+	}
+
+	if time.Since(ml.leaseStartTime) > lease {
+		ml.locked = false
+		return errors.New("lease lost")
+	}
+
+	ml.leaseStartTime = time.Now()
+	ml.renewsCounter += 1
+	return nil
+}
+
+type mockService struct {
+	startsCounter int
+	starterID     string
+}
+
+func (ms *mockService) Run(callerID string, ctx context.Context) func(ctx context.Context) {
+	return func(ctx context.Context) {
+
+		ms.startsCounter += 1
+		ms.starterID = callerID
+
+		<-ctx.Done()
+		ms.starterID = ""
+	}
+}
+
+func TestAcquireLock_MultipleInstances(t *testing.T) {
+	l := mockLock{
+		acquiresCalls: map[string]int{},
+	}
+
+	s := mockService{}
+
+	testCtx, testCancel := context.WithCancel(context.Background())
+	defer testCancel()
+
+	// Set up independent contexts to test the switch when one controller stops
+	hac1Ctx, hac1Cancel := context.WithCancel(testCtx)
+	defer hac1Cancel()
+
+	// Wait time on hac1 is 0, it should always get the lock.
+	hac1 := HAController{
+		ID:            "hac1",
+		lock:          &l,
+		logger:        testlog.HCLogger(t),
+		renewalPeriod: time.Duration(float64(lease) * renewalFactor),
+		waitPeriod:    time.Duration(float64(lease) * waitFactor),
+		randomDelay:   0,
+	}
+
+	hac2 := HAController{
+		ID:            "hac2",
+		lock:          &l,
+		logger:        testlog.HCLogger(t),
+		renewalPeriod: time.Duration(float64(lease) * renewalFactor),
+		waitPeriod:    time.Duration(float64(lease) * waitFactor),
+		randomDelay:   6 * time.Millisecond,
+	}
+
+	must.False(t, l.locked)
+
+	go hac1.Start(hac1Ctx, s.Run(hac1.ID, hac1Ctx))
+	go hac2.Start(testCtx, s.Run(hac2.ID, testCtx))
+
+	time.Sleep(4 * time.Millisecond)
+	/*
+		After 4 ms more (4 ms total):
+		* hac2 should  not have tried to acquire the lock.
+		* hac1 should have the lock and the service should be running.
+		* The first lease is not over yet, no calls to renew should have been made.
+	*/
+	must.True(t, l.locked)
+	must.Eq(t, 1, l.acquiresCalls[hac1.ID])
+	must.Eq(t, 0, l.acquiresCalls[hac2.ID])
+
+	must.Eq(t, 0, l.renewsCounter)
+
+	must.Eq(t, 1, s.startsCounter)
+	must.StrContains(t, hac1.ID, s.starterID)
+
+	time.Sleep(6 * time.Millisecond)
+	/*
+		After 6 ms more (10 ms total):
+		* hac2 should have tried to acquire the lock at least once.
```

Comment:
This is because the random delay is over right?

Reply:
yes!